### PR TITLE
make implicit team chat pass all client tests, and get CLI working CORE-6130

### PR DESCRIPTION
**Pull request by mmaxim**
Patch does the following:

1.) Gets all the chat tests to pass in `IMPTEAM` mode.
2.) Makes CLI work properly with `IMPTEAM` chats.

### CODE REVIEW

**Comment by mmaxim** [comment]
Needs https://github.com/keybase/keybase/pull/1766 to pass CI.

File: go/chat/tlf.go [hunk 1]
```diff
@@ -107,12 +108,9 @@ func (t *KBFSNameInfoSource) CryptKeys(ctx context.Context, tlfName string) (res
 	if runIdentify {
 		t.Debug(ectx, "CryptKeys: running identify")
 		group.Go(func() error {
-			query := keybase1.TLFQuery{
-				TlfName:          tlfName,
-				IdentifyBehavior: identBehavior,
-			}
 			var err error
-			ib, err = t.Identify(ectx, query, true)
+			names := strings.Split(strings.Fields(tlfName)[0], ",")
```
Comment (patrickxb) [inline go/chat/tlf.go]: This little guy could be a function.  It's repeated several times.  In convsource.go too.

**Comment by mmaxim** [comment]
@mlsteele Could you also take a look? Especially at the `teams` package.

File: go/client/chat_resolver.go [hunk 2]
```diff
@@ -45,41 +45,71 @@ type chatConversationResolvingBehavior struct {
...
+			return fmt.Errorf("failed to open chat conversation: %v", err)
+		}
+		if string(cname.CanonicalName) != tlfName {
+			// If we auto-complete TLF name, we should let users know.
+			// TODO: don't spam user here if it's just re-ordering
+			r.G.UI.GetTerminalUI().Printf("Using %s conversation %s.\n", visout, cname.CanonicalName)
+		}
+		req.ctx.canonicalizedTlfName = string(cname.CanonicalName)
+	case chat1.ConversationMembersType_IMPTEAM:
+		impRes, err := r.TeamsClient.LookupOrCreateImplicitTeam(ctx, keybase1.LookupOrCreateImplicitTeamArg{
```
Comment (mlsteele) [inline go/client/chat_resolver.go]: You want `OrCreate` here?

Comment (mmaxim) [reply go/client/chat_resolver.go]: Yes, if I send to `mikem,mlsteele` and it doesn't exist yet, I want it to be created. This is the same behavior as KBFS.

File: go/service/teams.go [hunk 3]
```diff
@@ -249,17 +249,38 @@ func (h *TeamsHandler) GetTeamRootID(ctx context.Context, id keybase1.TeamID) (k
...
+	}
+	return name, nil
+}
+
+func (h *TeamsHandler) LookupImplicitTeam(ctx context.Context, arg keybase1.LookupImplicitTeamArg) (res keybase1.LookupImplicitTeamRes, err error) {
 	defer h.G().CTraceTimed(ctx, fmt.Sprintf("LookupImplicitTeam(%s)", arg.Name), func() error { return err })()
-	teamID, _, err := teams.LookupImplicitTeam(ctx, h.G().ExternalG(), arg.Name, arg.Public)
-	return teamID, err
+
+	if arg.Name, err = h.prependUsername(ctx, arg.Name, arg.Public); err != nil {
```
Comment (mlsteele) [inline go/service/teams.go]: Why is this necessary? It doesn't make sense for teams you are a reader of like "other#you".

Comment (mmaxim) [reply go/service/teams.go]: Hmm true, I can move this up a level.

Comment (patrickxb) [reply go/service/teams.go]: same problem exists in kbfs chat i believe

Comment (mmaxim) [reply go/service/teams.go]: At this level it would affect actual KBFS though.

File: go/teams/loader.go [hunk 4]
```diff
@@ -382,7 +384,7 @@ func (l *TeamLoader) load2Inner(ctx context.Context, arg load2ArgT) (*load2ResT,
 			ret.Secretless = true
 		} else {
 			// Add the secrets unless this is a secretless team.
-			if !ret.Secretless {
+			if !ret.Secretless && !arg.public {
```
Comment (mlsteele) [inline go/teams/loader.go]: I think `!ret.Secretless && !ret.Chain.Public` would be a little nicer. Doesn't really matter

Comment (mmaxim) [reply go/teams/loader.go]: If they aren't the same, it's an error anyway.

Comment (mlsteele) [reply go/teams/loader.go]: yep but if they mismatch it will error here which would be harder to debug

**Comment by mlsteele** [comment]
@maxtaco does the server check whether seqtype is correct on incoming links? Apparently this is the first client PR to write them in and it would be too bad to have wrong ones in there.

**Comment by maxtaco** [comment]
Good q. Yes and no.  There's a ticket in the upcoming sprint to further investigate all combinations of error here.  I think @mmaxim was able to post something broken that surprised me, so I want to add more checks and tests around this.

**Comment by mmaxim** [comment]
I'm fine with punting on SBS and such for public chats for now if you all are.

**Comment by mlsteele** [comment]
It's not just SBS, it's CLKR too. How about making it only skip secrets if the team chain says you're not a member. Actually what motivated you to turn off secrets anyway? I imagine it's necessary for when you're not in the team but are loading it.

**Comment by mmaxim** [comment]
We just use the 0 key for public chats, so we never attempt to load them.

File: go/chat/teams.go [hunk 5]
```diff
@@ -80,22 +80,20 @@ func (t *ImplicitTeamsNameInfoSource) Lookup(ctx context.Context, name string, v
...
+		vis == keybase1.TLFVisibility_PUBLIC)
 	if err != nil {
 		return res, err
 	}
-
 	if !teamID.IsRootTeam() {
 		panic(fmt.Sprintf("implicit team found via LookupImplicitTeam not root team: %s", teamID))
 	}
 
+	res.CanonicalName = teamName.String()
```
Comment (mlsteele) [inline go/chat/teams.go]: It seems like letting the __ugly_name float around out here in chat is an unnecessary wart. It's annoying to have so many identifiers, and the __ugly_name only exists as a consequence of the fact that iteams happen to be built on teams. Could you use the team ID as the canonical name?

Sorry for stirring up so much ruckus on this PR.

**Comment by mlsteele** [comment]
You must load the team in the process of verifying/unboxing messages?

File: go/chat/teams.go (hunk 5 above)
Comment (mmaxim) [reply go/chat/teams.go]: This name gets written into the TLF name field of messages, it isn't displayed to anyone. I assume the name of this team needs to be represented somewhere on the message.

**Comment by mmaxim** [comment]
The load works because of the change I made. I thought you meant calling `ApplicationKeys`.

File: go/chat/teams.go (hunk 5 above)
Comment (mlsteele) [reply go/chat/teams.go]: Why would the __ugly_name need to be represented somewhere? I feel like it should be avoided.

Comment (mmaxim) [reply go/chat/teams.go]: Is there any substantive difference here? If we aren't going to display it anywhere, than what difference does it make if it is this name vs the ID? Is there something intrinsically nicer about an ID? At the bare minimum it would be nice to see the name when debugging, it may have a use. I already have the ID represented on the message.

Comment (mlsteele) [reply go/chat/teams.go]: I think it's a footgun to have multiple IDs and ways of addressing objects that need to be checked converted. But no there's no substantive difference. Not sure why it's nice to see for debugging.

File: go/chat/keyfinder.go [hunk 6]
```diff
@@ -131,7 +137,13 @@ func (k *KeyFinderImpl) FindForDecryption(ctx context.Context,
...
+	case chat1.ConversationMembersType_IMPTEAM:
+		if public {
+			// If this is public, then just make sure we get to the name info source
+			return k.Find(ctx, tlfName, membersType, public)
+		}
+		fallthrough
+	case chat1.ConversationMembersType_TEAM:
 		teamID, err := tlfIDToTeamdID(teamID)
 		if err != nil {
 			return res, err
```
Comment (mlsteele) [inline go/chat/keyfinder.go]: pass `public` to the `teams.Load` below. Might help avoid bugs in the future implementation of public non-implicit teams

Comment (mmaxim) [reply go/chat/keyfinder.go]: Sure
